Reducing AR::ConPool's critical (synchronized) section

thedarkone opened the pull request:
This is an alternative (much lighter in scope) PR to #14926, only the smallest possible amount of code is changed.

The objective is to:
- move connection verification outside of `synchronized` section -> a3923e6
- move connection establishment outside of the `synchronized` section -> e92f5a9

This should allow for more concurrency (gist with benched concurrency improvements: https://gist.github.com/thedarkone/80f78dea697f2a7f4150). Related: #14891, #14919.

As per @matthewd's suggestion:
- remove synchronization from around the `@reserved_connections` cache -> 603fe20

michaeldiscala commented:
Hi @thedarkone  -- do you plan to continue with this branch? I'm having a bit of trouble with rails and pgbouncer that I think is related to rails verifying the connection inside of a mutex.

I don't really know very much about connection pool's internals, but would be happy to pitch in if there's anything I can do to help move the branch along.

SimonKaluza commented:
:+1: 

These look great and my guess is they would fix quite a bit of the problems sidekiq users have been experiencing relating to ActiveRecord slowdowns/timeouts as discussed https://github.com/mperham/sidekiq/issues/921 and https://github.com/mperham/sidekiq/issues/702).

That said, I'm having a tough time trying to test these in my primary application due to the fact that this branch tracked 4.2alpha and I've had a tough time lining up my project's gem versions with those that were in development at the time.  I tried to merge these 4 commits into the latest version of 4.2.1, but I'm not sure how to reconcile conflicts between @thedarkone's latest additions and @al2o3cr's fixes for connection leaks in the already backported PR https://github.com/rails/rails/pull/18167.  Does this PR already account for the leaked connections?

thedarkone commented:
I'll rebase in a day or two.

SimonKaluza commented:
Nice!  Thanks @thedarkone :clap:

thedarkone committed:
AR::ConPool - reduce post checkout critical section.

Move post checkout connection verification out of mutex.synchronize.

thedarkone commented:
Rebased.

> Does this PR already account for the leaked connections?

Some of the changes make those leaks impossible, I've rolled back @al2o3cr's patch (while keeping his tests).

Not sure if this PR will have any effect on those sidekiq issues (linked above by @SimonKaluza).

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:
```diff
@@ -193,6 +205,80 @@ def wait_poll(timeout)
+      # Adds the ability to turn a basic fair FIFO queue into one
+      # biased to some thread.
+      module BiasableQueue # :nodoc:
+        class BiasedConditionVariable
```
`# :nodoc:`

tenderlove replied:
Is this something we could upstream to concurrent-ruby, and reduce the patch size? /cc @jdantonio

jdantonio replied:
@tenderlove Absolutely. @pitr-ch has done excellent work creating an enhanced synchronization layer and is currently working on an [improved condition variable](https://github.com/ruby-concurrency/concurrent-ruby/blob/synchronization/lib/concurrent/synchronization/condition.rb). I don't see why we couldn't add this. @pitr-ch: Thoughts?

pitr-ch replied:
@jdantonio thanks for cc-ing me. I'll have some time on weekend, I'll look in more detail.

pitr-ch replied:
Sorry, I've commented earlier but it looks like it never reached Github. I've pulled the rails source and tried to understand this class. I have to admit I was not able to grasp it. I know what Java uses "Biased locking" to optimize. I assume this is related but I am not sure about the conditions when this class can be used etc. @thedarkone could you provide more documentation for this class please.

tenderlove commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:
```diff
@@ -299,32 +404,82 @@ def connected?
+            end
+            @connections.delete_if do |conn|
+              conn.requires_reloading?
+            end
```
I think this can be `@connections.delete_if(&:requires_reloading?)`

tenderlove commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:
```diff
@@ -403,7 +570,104 @@ def reap
+            end
+          end
+        end
+      rescue Exception => e # if something went wrong
```
I'm not comfortable rescuing `Exception`.  Can this be rewritten to use `ExclusiveConnectionTimeoutError` and an empty `rescue`?  Or is there some reason we must rescue from `Exception`?

matthewd replied:
@tenderlove we have to do the checkin before we re-raise, for all abnormal exits (see e.g., https://github.com/rails/rails/issues/20114#issuecomment-101092729)

Maybe this would be clearer as two separate rescues though, despite their bodily similarity.

tenderlove replied:
I see what it's doing now.  Yes, please separate this in to two `rescue` blocks.  I'd rather have duplicate code than try to decipher the conditional below.

tenderlove commented:
Would it be possible to make a throughput test?  The PR looks good to me, but getting some numbers on how this improves throughput would be really awesome.

thedarkone committed:
AR::ConPool - establish connections outside of critical section.

thedarkone committed:
AR::ConPool - remove synchronization around connection cache.

Renamed `@reserved_connections` -> `@thread_cached_conns`. New name
clearly conveys the purpose of the cache, which is to speed-up
`#connection` method.

The new `@thread_cached_conns` now also uses `Thread` objects as keys
(instead of previously `Thread.current.object_id`).

Since there is no longer any synchronization around
`@thread_cached_conns`, `disconnect!` and `clear_reloadable_connections!`
methods now pre-emptively obtain ownership (via `checkout`) of all
existing connections, before modifying internal data structures.

A private method `release` has been renamed `thread_conn_uncache` to
clear-up its purpose.

Fixed some brittle `thread.status == "sleep"` tests (threads can go
into sleep even without locks).

thedarkone commented:
Re-pushed (with suggested changes applied).

> Would it be possible to make a throughput test? The PR looks good to me, but getting some numbers on how this improves throughput would be really awesome.

There's https://gist.github.com/thedarkone/80f78dea697f2a7f4150, it "benches"/verifies that connections are established and verified concurrently.

I've also attempted to bench a fake "typical" Rails request on JRuby HEAD: https://gist.github.com/thedarkone/e58974b8d81c9fd2a134. On my 4 core (8 hyperthread threads):
 * with a pool size of 8 and 10 concurrent requests (to simulate some contention), I've measured something like a 30% throughput improvement,
 * with a pool size of 8 and 7 concurrent requests (to simulate no contention), I've measured something like a 45% throughput improvement.

PS: Interestingly, running the throughput bench on Rails master I see about 135% jruby CPU usage, while with the patch applied, it results in 310% jruby CPU usage for the contented scenario (pool size: 8, threads: 10) and 250% jruby CPU usage for non contented (pool size: 8, threads: 7).

kamipo commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:
```diff
@@ -406,6 +449,43 @@ def reap
+        end
+        yield
+      ensure
+        synchronize { @new_cons_enabled = previous_value }
```
Looks like that `with_new_connections_blocked` have a race condition. This method sometimes causes `@new_cons_enabled = false` then never toggled to true.

meetme2meat commented:
@matthewd, @thedarkone  Can this be backported to 4.2.stable.

rafaelfranca commented:
@meetme2meat 4.2 only receives severe security fixes
